### ISSUE TITLE
IMA: Support midrolls in live streams

### ISSUE THREAD

**Issue by fpittersacc** [google/ExoPlayer#6388]
We are testing Live Stream with Midrolls and it seems Exoplayer gets stuck on an invalid state when Ads Uri Requests contains Midrolls.

As mentioned here (https://github.com/google/ExoPlayer/issues/5005#issuecomment-440601762) it seems Midrolls are not supported by Exoplayer IMA extension. 
Could you confirm it or if there is any plan to support it?

To reproduce it, use IMA Exo Demo and Ads Request with Midroll:
Content: http://livesim.dashif.org/livesim/testpic_2s/Manifest.mpd
Ads: https://pubads.g.doubleclick.net/gampad/ads?sz=640x480&iu=/124319096/external/ad_rule_samples&ciu_szs=300x250&ad_rule=1&impl=s&gdfp_req=1&env=vp&output=vmap&unviewed_position_start=1&cust_params=deployment%3Ddevsite%26sample_ar%3Dpremidpost&cmsid=496&vid=short_onecue&correlator=

**Comment by ojw28** [comment]
I don't think so, but @andrewlewis can confirm. I think we see server side ad insertion as by far the more common option for live streams.

**Comment by andrewlewis** [comment]
That's correct, they are not supported. (Prerolls should work with live streams, though.)

When resuming content after playing a midroll ad, where would you want playback to resume relative to the previously playing content position?

I will mark this as a low priority enhancement for now. Server-side ad insertion is probably a more popular option for this use case.

**Comment by okycelt** [comment]
@andrewlewis, as #7322 has been closed, should we re-send the testing stream and VMAP to this issue? 

Also, just so we know, when approximately do you think you will be able to allocate some time for this?

Thank you

**Comment by andrewlewis** [comment]
I was planning to look it up on the other issue but feel free to repost here. Thanks. Not sure yet when I will get round to this but it's not on our roadmap for the short term.

**Comment by aloj22** [comment]
Hi @andrewlewis , any update on this?

**Comment by marcbaechinger** [comment]
The IMA extension supports ads in HLA and DASH live stream only for DAI. There is currently no support for seeking, but I think the approach for live streams with IMA is DAI rather than client side.

Can you educate me around what you can do with the IMA SDK directly? Does the IMA SDK (without ExoPlayer extension) support client side ads in live streams?

**Comment by aloj22** [comment]
Hi @marcbaechinger, I haven't used IMA SDK standalone but ExoPlayer IMA extension. The problem is that if a midroll ad is set in live streaming and `playAdBeforeStartPosition` is set to true, the midroll is played instantly when starting playing because midroll ad start time is lesser than current playing time (which is current unix time in live streaming). 

Maybe I'm wrong with that assumption but that's what I think it's going on.

This is log I found: `AdTagLoader: Content progress: 22013620 ms of -1 ms` which is around 6 hours instead the current playing time

**Comment by marcbaechinger** [comment]
Ok, thanks. 

As stated above, live streams only support pre-rolls with the extension for client-side ads. Under the hood it is using the IMA SDK. 

The IMA SDK supports mid-roll with live streams for DAI ads. This is supported for HLS and DASH (except seeking). Please see https://github.com/google/ExoPlayer/issues/10912 also.
